packages/services/src/competition.service.ts

```diff
@@ -2338,6 +2356,82 @@ export class CompetitionService {
     );
   }
 
+  /**
+   * Validate agent against competition participation rules
+   * @param competition Competition with participation rules
+   * @param agentId Agent ID to validate
+   * @returns void (throws ApiError if validation fails)
+   */
+  private async validateParticipationRules(
+    competition: SelectCompetition,
+    agentId: string,
+  ): Promise<void> {
+    // Rule 1: Blocklist check (highest priority - reject immediately)
+    if (competition.blocklist && competition.blocklist.length > 0) {
+      if (competition.blocklist.includes(agentId)) {
+        throw new ApiError(
+          403,
+          "This agent is not permitted to join this competition",
+        );
+      }
+    }
+
+    // Rule 2: Allowlist-only mode
+    if (competition.allowlistOnly) {
+      if (!competition.allowlist || competition.allowlist.length === 0) {
+        // If allowlistOnly is true but no allowlist exists, nobody can join
+        throw new ApiError(
+          403,
+          "This competition is allowlist-only. Your agent is not on the allowlist",
```

Comment:
is there a way to let us know that the allowlist is not configured which is likely a mistake on our end, otherwise this error can be confused with below.